python/cudf/cudf/utils/_numba.py

```diff
@@ -0,0 +1,196 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+import glob
+import math
+import os
+import subprocess
+import sys
+import warnings
+
+from numba import config
+
+CC_60_PTX_FILE = os.path.dirname(__file__) + "/../core/udf/shim_60.ptx"
+_NO_DRIVER = (math.inf, math.inf)
+
+CMD = """\
+from ctypes import c_int, byref
+from numba import cuda
+dv = c_int(0)
+cuda.cudadrv.driver.driver.cuDriverGetVersion(byref(dv))
+drv_major = dv.value // 1000
+drv_minor = (dv.value - (drv_major * 1000)) // 10
+run_major, run_minor = cuda.runtime.get_version()
+print(f'{drv_major} {drv_minor} {run_major} {run_minor}')
+"""
+
+
+def _get_versions():
+    """
+    This function is mostly vendored from ptxcompiler and is used
+    to check the system CUDA driver and runtime versions in its absence.
+    """
+    cp = subprocess.run([sys.executable, "-c", CMD], capture_output=True)
```

Comment:
Does ptxcompiler use a subprocess currently? I am surprised by this because this tends to be costly. If possible, we should avoid launching a subprocess at cudf import time.

Reply:
Is there a way to use any other tool (cuda-python, rmm, ...) to know this information? I assume the goal is to avoid importing `numba.cuda` in this process...

Reply:
Yes, `ptxcompiler` uses a subprocess. This command is vendored directly from it [here](https://github.com/rapidsai/ptxcompiler/blob/main/ptxcompiler/patch.py#L136-L145). While doing all of our setup before importing `numba.cuda` is _one_ goal, another requirement is that the process for getting the versions doesn't `cuinit`, otherwise it can interfere with the way dask initializes its networking.

Reply:
As an aside it would be nice if we had some kind of small independent package we could list as a dependency that served as a source of truth for the driver and runtime versions, which we could also configure to avoid forking a subprocess, calling cuinit, etc. This would avoid a bunch of reinventing the wheel across cuda-python, numba, rmm, etc....

Reply:
I tried putting a 
```python
from cuda import cuda, cudart
cuda.cuDriverGetVersion()
cudart.cudaRuntimeGetVersion()
```
At the top of `cudf`'s `__init__.py` and ran `test_no_cuinit.py`, which passed. So I suppose it's a clean approach from the `cuInit` perspective. Happy to refactor as such here and see if things work.

Reply:
Think this is referring to issue ( https://github.com/NVIDIA/cuda-python/issues/16 )

Namely the version cuda-python returns is hard-coded at build time as opposed to querying it at run time. So the version it returns may not reflect what the user has installed on their system (and will instead reflect how the binary was built). The issue explains why this is happening currently (though it still presents an issue with intended usage)

There's more details in the threads in PR ( https://github.com/rapidsai/rmm/pull/946 ) as well

Reply:
To close the loop on this I decided in https://github.com/rapidsai/cudf/pull/13337/commits/5cb0ce6d67b2f2c8813f3140cbbf3311ca4c7743 to just vendor the three functions to `_ptxcompiler.py`. 

To summarize the way it works now, the user either has `ptxcompiler` or doesn't. If they do, use it to get the versions (this launches a subprocess). If they don't, use the vendored functions to do so (this also launches the same subprocess using the same command). 

We are constrained to use a subprocess for three reasons:
1. We can not call `cuinit` on `cudf`'s import without interfering with dask. 
2. We can not directly use `numba.cuda` to obtain the versions because we need to finish configuring numba before `numba.cuda` is imported.
3. We can not utilize `cuda.cuda` due to constraint described above

Even _with_ the approach above we need to leave in place a mechanism to _disable_ the spawning of a subprocess in HPC environments where it is not safe to do so. In this case the user may provide an environment variable which disables the check and two more environment variables specifying the driver and runtime versions manually, and these are parsed and returned instead. These environment variables are the same ones, between `ptxcompiler` and our vendored version.